sms_statistics_service/sms_statistics_service.py

```diff
@@ -0,0 +1,100 @@
+                             "the operations statistics")
+    parser.add_argument("start_minute_inclusive", metavar="start-minute-inclusive",
+                        help="ISO 8601 string for the start of the datetime range to update sms statistics for")
+    parser.add_argument("end_minute_exclusive", metavar="end-minute-inclusive",
+                        help="ISO 8601 string for the end of the datetime range to update sms statistics for ")
+
+    args = parser.parse_args()
+
+    google_cloud_credentials_file_path = args.google_cloud_credentials_file_path
+    firestore_credentials_url = args.firestore_credentials_url
+    start_minute_inclusive = isoparse(args.start_minute_inclusive)
+    end_minute_exclusive = isoparse(args.end_minute_exclusive)
+
+    assert start_minute_inclusive.second == 0
+    assert start_minute_inclusive.microsecond == 0
+    assert end_minute_exclusive.second == 0
+    assert end_minute_exclusive.microsecond == 0
+
+    log.info("Initialising the Firestore client...")
+    firestore_credentials = json.loads(google_cloud_utils.download_blob_to_string(
+            google_cloud_credentials_file_path, firestore_credentials_url))
+    firestore_client = FirestoreClient(firestore_credentials)
+
+    log.info("Downloading the active project details from Firestore...")
+    active_projects = firestore_client.get_active_projects()
+    log.info(f"Downloaded the details for {len(active_projects)} active projects")
+
+    for project in active_projects:
+        log.info(f"Computing SMS statistics for project {project.project_name}...")
+
+        log.info(f"Downloading the Rapid Pro credentials token from '{project.rapid_pro_token_url}'...")
+        rapid_pro_token = google_cloud_utils.download_blob_to_string(
+            google_cloud_credentials_file_path, project.rapid_pro_token_url).strip()
+
+        log.info(f"Downloading raw messages from Rapid Pro...")
+        rapid_pro = RapidProClient(project.rapid_pro_domain, rapid_pro_token)
+        raw_messages = rapid_pro.get_raw_messages(created_after_inclusive=start_minute_inclusive,
+                                                  created_before_exclusive=end_minute_exclusive)
+
+        log.info("Computing message stats for each minute...")
+        # Create a table of counts for all the minutes of interest, with all counts initialised to 0
+        stats = dict()  # of minute iso_string -> SMSStats
+        minute = start_minute_inclusive
+        while minute < end_minute_exclusive:
+            stats[minute.astimezone(pytz.utc).isoformat(timespec="minutes")] = SMSStats()
+            minute += datetime.timedelta(minutes=1)
+
+        # Loop over all of the downloaded messages and increment the appropriate count
+        unhandled_status_count = 0
+        for msg in raw_messages:
+            minute_stats = stats[msg.created_on.astimezone(pytz.utc).isoformat(timespec="minutes")]
+
+            if msg.direction == "in":
```

Comment:
Consider using continue to decrease the nesting of this.

```
if ...:
   continue

if direction != out:
   log ("unexpected direction")

assert direction == 'out'
if msg.status == ....
```

Reply:
changed